### DISCUSSION TITLE
Is there a recommended way to handle the form editing scenario in Jotai?

### DISCUSSION THREAD

**Discussion by yuezk** [pmndrs/jotai#2022] [Q&A] [answered]
Jotai version: 2.2.1

Scenario: form editing
1. Fetch the user info via the HTTP api call to fill the form
2. Edit the user info
3. Click Save to save the updated user info to the server side

Problem: it always shows the fallback UI of the Suspense component
whenever the form fields change

Question: is there a recommended way to handle this scenario in Jotai? i.e., avoid always triggering the suspense UI when performing the local editing.

Demo: https://codesandbox.io/s/bold-framework-6ss6qq?file=/src/App.tsx

Similar items: https://github.com/pmndrs/jotai/discussions/1458, https://github.com/pmndrs/jotai/discussions/2003, https://github.com/pmndrs/jotai/issues/1007

**Comment by yuezk** [comment]
It looks like I can use the `loadable` atom to avoid the suspense. See https://codesandbox.io/s/crazy-grass-gqf5wt

Not sure whether it is the preferred way.

**Reply by dai-shi** [reply]
Yeah, `loadable` would rescue. In this case, `unstable_wrap` fits better.

Alternatively, something like this would work, but it's not in sync and the input focus is gone.
https://codesandbox.io/s/elated-sunset-j3dlrx

**Reply by yuezk** [reply]
Inspired by your code, I made the type of `firstNameAtom` and the `lastNameAtom` as `string | Promise<string>`. Then I could refactor the code as https://codesandbox.io/s/distracted-meadow-h396sw?file=/src/App.tsx, the downside is that I didn't find a way to sync the user-modified data into the `userAtom`, but it doesn't matter for me. Wondering if that trick is acceptable.

**Reply by dai-shi** [reply, marked as answer]
It looks very good, and it's atomic.

You could do this:
```js
const initialUserAtom = atom(async () => ...);
const firstNameAtom = atomWithDefault(...);
const lastNameAtom = atomWithDefault(...);
const currentUserAtom = atom((get) => ({ firstName: get(firstNameAtom), lastName: get(lastNameAtom) }));
```

**Reply by yuezk** [reply]
Thanks for your advice. Looks clear. My question regarding the `currentUserAtom` is that. I can only use it when submitting the data to the server, but I should avoid using it in the React component. Otherwise, it will trigger the Suspense, is that correct? 

```typescript
const currentUserAtom = atom<Promise<User>>(async get => {
  const firstName = await get(firstNameAtom)
  const lastName = await get(lastNameAtom)
  return {
    firstName,
    lastName
  }
})
```

**Reply by dai-shi** [reply]
Oh, I forgot that firstNameAtom can be async at first. Your version seems correct.

To avoid suspense for async atoms, there are some workarounds, but you probably want to try `unstable_unwrap` which isn't documented yet, but types should probably tell.

**Reply by yuezk** [reply]
Okay, let me wait for the stabilization of `unstable_unwrap`, before that, the workaround is just sufficient for me. Thanks.
